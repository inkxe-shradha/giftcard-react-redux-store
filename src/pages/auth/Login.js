import {
  Avatar,
  Container,
  Grid,
  Paper,
  TextField,
  Typography,
  Alert,
  IconButton,
} from "@mui/material";
import { makeStyles } from "@mui/styles";
import LockOutlinedIcon from "@mui/icons-material/LockOutlined";
import React, { useEffect } from "react";
import { useFormik } from "formik";
import * as yup from "yup";
import { useLocation, useNavigate } from "react-router-dom";
import KeyboardArrowRight from "@mui/icons-material/KeyboardArrowRight";
import LoadingButton from "@mui/lab/LoadingButton";
import GoogleLogin from "react-google-login";
import ArrowBackIcon from "@mui/icons-material/ArrowBack";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { signInUser } from "../../store/actions/authAction";
import { toast } from "react-toastify";
const useStyles = makeStyles((theme) => ({
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary?.main,
  },
  margin: {
    margin: theme.spacing(3),
  },
  padding: {
    padding: theme.spacing(2),
  },
  paper: {
    margin: theme.spacing(3),
    padding: theme.spacing(2),
  },
  google: {
    borderRadius: theme.spacing(2),
    margin: theme.spacing(1),
  },
  container: {
    display: "flex !important",
    flexWrap: "wrap",
    alignItems: "center",
    justifyContent: "center",
    height: "100vh",
  },
  backButton: {
    float: "right",
  },
}));

const Login = ({ loading, signInUser, isAuthenticated, demoCall }) => {
  const classes = useStyles();
  const { state } = useLocation();
  const navigate = useNavigate();

  useEffect(() => {
    if (isAuthenticated) {
      navigate(state?.path || "/");
    }
  }, [isAuthenticated]);
  const validationSchema = yup.object({
    email: yup
      .string("Enter your email")
      .email("Enter a valid email")
      .required("Email is required"),
    password: yup
      .string("Enter your password")
      .min(8, "Password should be of minimum 8 characters length")
      .required("Password is required"),
  });

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validationSchema,
    onSubmit: (values) => {
      const signUpObj = {
        email: values.email,
        name: "Admin",
        imageUrl: "https://source.unsplash.com/random/900X700/?avatar",
        balance_points: 10000,
        wishlist: [],
        cards_gifted: [],
        cards_received: [],
        role: "admin",
      };
      signInUser(signUpObj);
    },
  });
  // Google Validation Schemas
  const responseGoogle = (response) => {
    const signUpObj = {
      email: response.profileObj.email,
      name: response.profileObj.name,
      imageUrl: response.profileObj.imageUrl,
      balance_points: 5000,
      wishlist: [],
      cards_gifted: [],
      cards_received: [],
      role: "user",
      id: response.profileObj.googleId,
    };
    signInUser(signUpObj, "google");
  };

  const handelError = (error) => {
    toast.success(
      "Failed to logged in! Please tyr again with your google account"
    );
  };
  return (
    <Container maxWidth="sm" component="main" className={classes.container}>
      <Paper className={classes.paper} elevation={2}>
        <Grid container spacing={2} align="center" justifyContent="center">
          <Grid item xs={12} sm={12} md={12}>
            {/* Back button */}
            <IconButton
              color="default"
              aria-label="back button"
              className={classes.backButton}
              onClick={() => navigate("/")}
            >
              <ArrowBackIcon />
            </IconButton>
            <Avatar className={classes.avatar}>
              <LockOutlinedIcon />
            </Avatar>
            <Typography variant="h4">Login</Typography>
          </Grid>
          <Grid item xs={12}>
            <form
              onSubmit={formik.handleSubmit}
              className={classes.form}
              noValidate
            >
              {/* {errorMess && <Alert severity="error">{errorMess}</Alert>} */}
              <TextField
                variant="outlined"
                margin="normal"
                required
                fullWidth
                placeholder="User Email Id"
                id="email"
                name="email"
                label="Email"
                value={formik.values.email}
                onChange={formik.handleChange}
                error={formik.touched.email && Boolean(formik.errors.email)}
                helperText={formik.touched.email && formik.errors.email}
                autoFocus
              />
              <TextField
                variant="outlined"
                margin="normal"
                required
                fullWidth
                id="password"
                name="password"
                label="Password"
                type="password"
                placeholder="*******"
                value={formik.values.password}
                onChange={formik.handleChange}
                error={
                  formik.touched.password && Boolean(formik.errors.password)
                }
                helperText={formik.touched.password && formik.errors.password}
                autoComplete="current-password"
              />
              <LoadingButton
                type="submit"
                fullWidth
                variant="contained"
                loading={loading}
                color="primary"
                endIcon={<KeyboardArrowRight />}
                className={classes.submit}
              >
                Log in
              </LoadingButton>
            </form>
            <hr className="hr-text" data-content="OR" />
            <GoogleLogin
              clientId="25860778518-lt39cmgjc3s1na7pfc564gdn2bilqljq.apps.googleusercontent.com"
              buttonText="Sign In With Google"
              className={classes.google}
              onSuccess={responseGoogle}
              autoLoad={false}
              onFailure={handelError}
              cookiePolicy={"single_host_origin"}
            />
          </Grid>
        </Grid>
      </Paper>
    </Container>
  );
};

const mapStateToProps = (state) => ({
  loading: state.loading.loading,
  isAuthenticated: state.auth.isAuthenticated,
});

const mapDispatchToProps = {
  signInUser,
};

// Prototypes
Login.propTypes = {
  loading: PropTypes.bool.isRequired,
  signInUser: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);
