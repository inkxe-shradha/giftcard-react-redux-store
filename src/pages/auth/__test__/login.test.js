import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Login from "../Login";
import thunk from "redux-thunk";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { ThemeProvider } from "@mui/system";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { createTheme } from "@mui/material";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const theme = createTheme();
const state = {
  auth: {
    user: {},
    isAuthenticated: false,
  },
  loading: {
    loading: false,
  },
};
const customRender = (component, location, initialState = state) => {
  const history = createMemoryHistory();
  history.push(location);
  const store = mockStore(initialState);
  return render(
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <Router location={history.location} navigator={history}>
          {component}
        </Router>
      </ThemeProvider>
    </Provider>
  );
};

describe("The authentication page test suite", () => {
  it("Should create a snapshot of the login  page", () => {
    const { asFragment } = customRender(<Login />, "/login");
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should sign in button present on the component", () => {
    customRender(<Login />, "/login");
    const linkElement = screen.getAllByText(/Sign In/i);
    expect(linkElement).toBeTruthy();
  });

  it("Input should be initially empty", () => {
    customRender(<Login />, "/login");
    const emailElement = screen.getByRole("textbox", { name: "Email" });
    const passwordElement = screen.getByLabelText(/password/i);
    expect(emailElement.value).toBe("");
    expect(passwordElement.value).toBe("");
  });

  it("Should be able to type an email", async () => {
    customRender(<Login />, "/login");
    const emailElement = screen.getByRole("textbox", { name: "Email" });
    fireEvent.change(emailElement, {
      target: { value: "shradhasuman2@gmail.com" },
    });
    await waitFor(() => {
      expect(emailElement.value).toBe("shradhasuman2@gmail.com");
    });
  });

  it("Should be able to type a password", async () => {
    customRender(<Login />, "/login");
    const passwordElement = screen.getByLabelText(/password/i);
    fireEvent.change(passwordElement, { target: { value: "12345678" } });
    await waitFor(() => {
      expect(passwordElement.value).toBe("12345678");
    });
  });

  it("Should be show invalid email if the email is not valid", async () => {
    customRender(<Login />, "/login");
    const errorElement = screen.queryByText(/Enter a valid email/i);
    expect(errorElement).not.toBeInTheDocument();
    // getting the email field for testing purposes
    const emailElement = screen.getByRole("textbox", { name: "Email" });
    fireEvent.change(emailElement, { target: { value: "shradhasuman.com" } });
    const submitButtonElement = screen.getByRole("button", { name: /Log in/i });

    fireEvent.click(submitButtonElement);
    await waitFor(() => {
      const emailErrorAfter = screen.getByText("Enter a valid email");
      expect(emailErrorAfter).toBeInTheDocument();
    });
  });

  it("Should be show invalid password if the password  length is less than 8", async () => {
    customRender(<Login />, "/login");
    const errorElement = screen.queryByText(
      /Password should be of minimum 8 characters length/i
    );
    expect(errorElement).not.toBeInTheDocument();
    // getting the email field for testing purposes
    const passwordElement = screen.getByLabelText(/password/i);
    fireEvent.change(passwordElement, { target: { value: "12345" } });
    const submitButtonElement = screen.getByRole("button", {
      name: /Log in/i,
    });

    fireEvent.click(submitButtonElement);
    await waitFor(() => {
      const passwordErrorAfter = screen.getByText(
        "Password should be of minimum 8 characters length"
      );
      expect(passwordErrorAfter).toBeInTheDocument();
    });
  });

  it("Should dispatch if the user given valid credentials", async () => {
    const history = createMemoryHistory();
    history.push("/login");
    customRender(<Login />, "/login");
    const emailElement = screen.getByRole("textbox", { name: "Email" });
    fireEvent.change(emailElement, { target: { value: "admin@gmail.com" } });
    const passwordElement = screen.getByLabelText(/password/i);
    fireEvent.change(passwordElement, { target: { value: "12345678" } });
    const submitButtonElement = screen.getByRole("button", {
      name: /Log in/i,
    });
    fireEvent.click(submitButtonElement);
    await waitFor(() => {
      const signInUser = jest.fn();
      const store = mockStore({});
      store.dispatch(signInUser);
      expect(signInUser).toHaveBeenCalled();
    });
  });
});
