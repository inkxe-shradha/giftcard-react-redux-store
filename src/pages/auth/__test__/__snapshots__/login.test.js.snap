// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The authentication page test suite Should create a snapshot of the login  page 1`] = `
<DocumentFragment>
  <main
    class="MuiContainer-root MuiContainer-maxWidthSm makeStyles-container-6 css-cuefkz-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 makeStyles-paper-4 css-1mbunh-MuiPaper-root"
    >
      <div
        align="center"
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-12n0uxm-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12 css-1idn90j-MuiGrid-root"
        >
          <button
            aria-label="back button"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium makeStyles-backButton-7 css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault makeStyles-avatar-1 css-2s90m6-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="LockOutlinedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
              />
            </svg>
          </div>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Login
          </h4>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
        >
          <form
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="email"
                id="email-label"
              >
                Email
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="email"
                  name="email"
                  placeholder="User Email Id"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1in441m"
                  >
                    <span>
                      Email *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="password"
                id="password-label"
              >
                Password
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="current-password"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="password"
                  name="password"
                  placeholder="*******"
                  required=""
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1ftyaf0"
                  >
                    <span>
                      Password *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root css-utm9vu-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
              id=":r2:"
              tabindex="0"
              type="submit"
            >
              Log in
              <span
                class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyboardArrowRightIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </form>
          <hr
            class="hr-text"
            data-content="OR"
          />
          <button
            class="makeStyles-google-5"
            disabled=""
            style="background-color: rgb(255, 255, 255); display: inline-flex; align-items: center; color: rgba(0, 0, 0, 0.54); box-shadow: 0 2px 2px 0 rgba(0, 0, 0, .24), 0 0 1px 0 rgba(0, 0, 0, .24); padding: 0px; border-radius: 2px; border: 1px solid transparent; font-size: 14px; font-weight: 500; font-family: Roboto, sans-serif; opacity: 0.6;"
            type="button"
          >
            <div
              style="margin-right: 10px; background: rgb(255, 255, 255); padding: 10px; border-radius: 2px;"
            >
              <svg
                height="18"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="#000"
                  fill-rule="evenodd"
                >
                  <path
                    d="M9 3.48c1.69 0 2.83.73 3.48 1.34l2.54-2.48C13.46.89 11.43 0 9 0 5.48 0 2.44 2.02.96 4.96l2.91 2.26C4.6 5.05 6.62 3.48 9 3.48z"
                    fill="#EA4335"
                  />
                  <path
                    d="M17.64 9.2c0-.74-.06-1.28-.19-1.84H9v3.34h4.96c-.1.83-.64 2.08-1.84 2.92l2.84 2.2c1.7-1.57 2.68-3.88 2.68-6.62z"
                    fill="#4285F4"
                  />
                  <path
                    d="M3.88 10.78A5.54 5.54 0 0 1 3.58 9c0-.62.11-1.22.29-1.78L.96 4.96A9.008 9.008 0 0 0 0 9c0 1.45.35 2.82.96 4.04l2.92-2.26z"
                    fill="#FBBC05"
                  />
                  <path
                    d="M9 18c2.43 0 4.47-.8 5.96-2.18l-2.84-2.2c-.76.53-1.78.9-3.12.9-2.38 0-4.4-1.57-5.12-3.74L.97 13.04C2.45 15.98 5.48 18 9 18z"
                    fill="#34A853"
                  />
                  <path
                    d="M0 0h18v18H0z"
                    fill="none"
                  />
                </g>
              </svg>
            </div>
            <span
              style="font-weight: 500; padding: 10px 10px 10px 0px;"
            >
              Sign In With Google
            </span>
          </button>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
